server/src/main/java/org/opensearch/index/ShardIndexingPressureStore.java

```diff
@@ -0,0 +1,115 @@
+    }
+
+    public ShardIndexingPressureTracker getShardIndexingPressureTracker(ShardId shardId) {
+        ShardIndexingPressureTracker tracker = shardIndexingPressureHotStore.get((long)shardId.hashCode());
+        if (isNull(tracker)) {
+            // Attempt from Indexing pressure cold store
+            tracker = shardIndexingPressureColdStore.get((long)shardId.hashCode());
+            // If not present in cold store so instantiate a new one
+            if (isNull(tracker)) {
+                ShardIndexingPressureTracker newShardIndexingPressureTracker = new ShardIndexingPressureTracker(shardId,
+                    this.shardIndexingPressureSettings.getShardPrimaryAndCoordinatingBaseLimits(),
+                    this.shardIndexingPressureSettings.getShardReplicaBaseLimits());
+                // Try update the new shard stat to the hot store
+                tracker = shardIndexingPressureHotStore.putIfAbsent((long) shardId.hashCode(), newShardIndexingPressureTracker);
+                // Update the tracker so that we use the one actual in the hot store
+                tracker = tracker == null ? newShardIndexingPressureTracker : tracker;
+                // Write through into the cold store for future reference
+                updateIndexingPressureColdStore(tracker);
+            } else {
+                // Attempt update tracker to the primary store and return tracker finally in the store to avoid any race
+                ShardIndexingPressureTracker newTracker = shardIndexingPressureHotStore.putIfAbsent((long) shardId.hashCode(), tracker);
+                tracker = newTracker == null ? tracker : newTracker;
+            }
+        }
+        return tracker;
+    }
+
+    public Map<Long, ShardIndexingPressureTracker> getShardIndexingPressureHotStore() {
+        return Collections.unmodifiableMap(shardIndexingPressureHotStore);
+    }
+
+    public Map<Long, ShardIndexingPressureTracker> getShardIndexingPressureColdStore() {
+        return Collections.unmodifiableMap(shardIndexingPressureColdStore);
+    }
+
+    public void tryIndexingPressureTrackerCleanup(ShardIndexingPressureTracker tracker) {
+        if (tracker.memory().getCurrentCombinedCoordinatingAndPrimaryBytes().get() == 0 &&
+            tracker.memory().getCurrentReplicaBytes().get() == 0) {
+            // Try inserting into cache again in case there was an eviction earlier
+            shardIndexingPressureColdStore.putIfAbsent((long)tracker.getShardId().hashCode(), tracker);
+            // Remove from the active store
+            shardIndexingPressureHotStore.remove((long)tracker.getShardId().hashCode(), tracker);
+        }
+    }
+
+    private void updateIndexingPressureColdStore(ShardIndexingPressureTracker tracker) {
+        if (shardIndexingPressureColdStore.size() > maxColdStoreSize) {
+            shardIndexingPressureColdStore.clear();
+        }
+        shardIndexingPressureColdStore.put((long)tracker.getShardId().hashCode(), tracker);
+    }
```

Comment:
Clearing the cold store just because it got full doesn't seem ideal. Maybe we can improve the data structure used here